### PR TITLE
Thread Safe Motor Controller Driver

### PR DESCRIPTION
**Pull request by manthanand**
Made motor controller thread safe. Removed redundant structs and added error functionality to driver. Created bitmap for motor error

### CODE REVIEW

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:28]:
```suggestion
static void assertTritiumError(uint16_t motor_err){
```

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:101]: I feel like the code for copying the floats into an array could be better. Maybe something like
```
uint8_t data[8];
memcpy(data, &motorCurrent, sizeof(motorCurrent));
memcpy(data + sizeof(motorCurrent), &newVeloctiy, sizeof(newVelocity));
```

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:181]: Personally, I prefer the newline and curly braces for if statements, even if there's just one line because I think it's easier to read

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:203]: This is already atomic. You don't need a mutex for this

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:236]: This is already atomic

File: BSP/STM32F413/Src/BSP_CAN.C
Comment (manthanand) [reply BSP/STM32F413/Src/BSP_CAN.C:1]: different branch of controls

File: Drivers/Src/MotorController.c
Comment (manthanand) [reply Drivers/Src/MotorController.c:203]: Should we leave it just in case we have an update thread and a read thread? Or should I just remove it anyways. I don't think it will hurt to leave it in there.

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:98]: You can delete `index`, `nv`, and `mc`, since they aren't used anymore

**Comment by babusid** [comment]
> why no build

It seems to not build properly on here, but does it build locally?

**Comment by babusid** [comment]
BSP_CAN.C was an incorrect file, it should be called BSP_CAN.c with a lowercase C. There are BSP changes here that are made redundant by the changes that are in SendCarCAN PR, and some assertion stuff was merged into master that's also in here. I'm going to hold off on merging this until I've merged RTOS_SendCarCAN and master is either merged into this or rebased off of.

File: Drivers/Src/MotorController.c
Comment (babusid) [inline Drivers/Src/MotorController.c:18]: We should initialize the CurrentVelocity to 0, because if we try to access it before a MotorController_Read gets a MOTOR_DRIVE message and updates it, we'll have undefined behavior

**Review by babusid — CHANGES_REQUESTED** [review]

**Review by babusid — APPROVED** [review]
Looks good to me. I'll merge it in once we get another pair of eyes on it.

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:19]: Would be more readable if this was initialized to `T_NONE` instead of `0` to be consistent with `assertTritiumError()`

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:176]: This part technically isn't thread-safe, since `Motor_FaultBitmap` is a global and you are modifying it non-atomically. I don't think it matters though unless you call this function from multiple threads and care about distinguishing between different motor errors (neither of which I think y'all are doing)

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:133]: Lines 133-154 could use some cleaning up, but it might not be the best use of our time.

For example, 
```C
        //get first number (bits 0-31)
        for(int j = 0; j < MAX_CAN_LEN/2; j++){
            firstSum += data[j];
            if(j != MAX_CAN_LEN/2 - 1){
                firstSum = firstSum << 8;
            }
        }
```
could be replaced by
```C
        //get first number (bits 0-31)
        for(int j = 0; j < MAX_CAN_LEN/2; j++){
            firstSum <<= 8;
            firstSum += data[j];
        }
```
and `union convert` doesn't look like it is used anywhere

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:13]: I think we should put these inside of parentheses or declare them as `const uint32_t`, so we don't get any nasty surprises when the preprocessor pastes these in and the `<<` operator has lower precedence than some other operator

File: Drivers/Src/MotorController.c
Comment (manthanand) [reply Drivers/Src/MotorController.c:134]: convert.f is used for velocity

Comment (ClarkPoon) [reply Drivers/Src/MotorController.c:134]: We could replace
```C
convert.n = secondSum;
CurrentVelocity = convert.f;
```
with
```C
memcpy(&CurrentVelocity, &secondSum, sizeof(CurrentVelocity));
```
or
```C
CurrentVelocity = *(float *) &secondSum;
```
and delete the union to make it more readable, but I'm fine with it if you leave it how it is.

**Review by ClarkPoon — DISMISSED** [review]

**Review by Cam0Cow — DISMISSED** [review]

**Review by Cam0Cow — DISMISSED** [review]

File: Drivers/Src/MotorController.c
Comment (babusid) [inline Drivers/Src/MotorController.c:130]: This indentation is kinda weird

File: Drivers/Src/MotorController.c
Comment (babusid) [reply Drivers/Src/MotorController.c:168]: I think it does matter because we have different behaviors in our fault thread for different motor faults, so I think it does matter, because we don't want a non-critical error to beat a critical error to the fault thread.

**Review by babusid — CHANGES_REQUESTED** [review]

File: Drivers/Src/MotorController.c
Comment (babusid) [inline Drivers/Src/MotorController.c:197]: I don't think we'll ever recieve a MOTOR_DRIVE command, we only send it. I don't see handling for the actual messages we receive as outlined in the tritium datasheet.

Comment (babusid) [reply Drivers/Src/MotorController.c:197]: We will be receiving a MOTOR_VELOCITY message, I think the case statement got messed up here.

**Review by Cam0Cow — DISMISSED** [review]

File: Drivers/Src/MotorController.c
Comment (babusid) [inline Drivers/Src/MotorController.c:157]: This doesn't return a length, BSP_CAN_Read returns an ERRORSTATUS enum value

File: Drivers/Inc/MotorController.h
Comment (ClarkPoon) [inline Drivers/Inc/MotorController.h:32]:
```suggestion
 * @param   busCurrentFractionalSetPoint fraction of the bus current to allow the motor to draw
```

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:111]: Do you want any parameter checking in this function? What happens if you get passed an invalid parameter (and is this possible)?

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:157]: Is it possible for this function to error out and not release the mailbox semaphore?

File: Drivers/Src/MotorController.c
Comment (babusid) [inline Drivers/Src/MotorController.c:217]: Not sure this is actually necessary, I think we'll end up doing a motor controller read every time and just using the values returned from that, but we can leave this in. Whoever writes updateVelocity may use this for convenience instead of manually unpacking the data.

**Review by babusid — DISMISSED** [review]

File: BSP/STM32F413/Src/BSP_CAN.c
Comment (ClarkPoon) [inline BSP/STM32F413/Src/BSP_CAN.c:135]: Do you still need to do this TODO?

File: BSP/STM32F413/Src/BSP_CAN.c
Comment (ClarkPoon) [inline BSP/STM32F413/Src/BSP_CAN.c:181]: Can we get rid of this TODO comment?

File: BSP/STM32F413/Src/BSP_CAN.c
Comment (ClarkPoon) [inline BSP/STM32F413/Src/BSP_CAN.c:223]: Do we still need to do this?

File: Tests/Test_MotorController.c
Comment (ClarkPoon) [inline Tests/Test_MotorController.c:51]: Is this comment accurate? If so, having `5.0f` correspond to 1 amp is a bad interface

File: Drivers/Src/MotorController.c
Comment (ClarkPoon) [inline Drivers/Src/MotorController.c:90]: Might be nice to have a comment or make this use `sizeof` instead of a magic 4

**Review by ClarkPoon — COMMENTED** [review]
I have some small suggestions for this. If y'all want to go ahead and merge with no more fixes, let me know

**Review by babusid — APPROVED** [review]

**Review by ClarkPoon — APPROVED** [review]

**Pull request merged**
